tesla_ver/bubble_chart.py

```diff
@@ -14,185 +18,166 @@ def generateBubbleChart(server):
-        date are just information paramteres
-        """
-        _, content_string = contents.split(",")
-
-        decoded = base64.b64decode(content_string)
-        fileish = io.StringIO(decoded.decode("utf-8"))
-        df = pd.read_csv(fileish)
-
-        return df
-
     @app.callback(
         Output("hidden-data", "children"),
-        [
-            Input("upload", "contents"),
-            Input("upload", "filename"),
-            Input("upload", "last_modified"),
-        ],
+        [Input("upload", "contents"), Input("upload", "filename"), Input("upload", "last_modified"),],
     )
-    def upload_data(contents, filename, last_modified):
+    def upload_data(list_of_contents, list_of_filenames, _):
         """This callback handles storing the dataframe as JSON in the hidden
         component."""
+
+        def parse_contents(contents, filenames):
+            """Parse a Dash Upload into a DataFrame.
+
+            contents is a string read from the upload component.
+            filename and date are just information paramteres
+            """
+            # Parses contents into dataframes
+            df_list = list()
+            filename_titles = list()
+            for content, filename in zip(contents, filenames):
+                _, content_string = content.split(",")
+                # Append filename
+                filename = Path(filename).stem
+                filename_titles.append(filename)
+                decoded = base64.b64decode(content_string)
+                fileish = io.StringIO(decoded.decode("utf-8"))
+                df = pd.read_csv(fileish, sep="\t")
+                # Separates time series into a single value
+                df[["X", "Y"]] = df[["X", "Y"]].applymap(lambda x: x.split(","))
+                df = df.apply(pd.Series.explode)
+                df.reset_index(drop=True, inplace=True)
+                df = df.rename(columns={"Y": filename})
+                df_list.append(df)
+
+            # Merges all dataframes into a single list
+            df = reduce(
+                lambda left_frame, right_frame: pd.merge_ordered(
+                    left_frame,
+                    right_frame,
+                    on=[column for column in left_frame.columns if column not in [*filename_titles]],
+                    left_by="X",
+                    fill_method="ffill",
+                ),
+                df_list,
+            )
+            df = df.convert_dtypes()
+            df[filename_titles] = df[filename_titles].apply(pd.to_numeric, errors="coerce")
+            df[df.select_dtypes(np.number).columns] = df[df.select_dtypes(np.number).columns].fillna(0)
+            return df
+
         df = None
-        if contents is not None:
-            df = parse_contents(contents).to_json()
-        return df
+        if list_of_contents is not None:
+            df = parse_contents(list_of_contents, list_of_filenames)
+            return df.to_json()
+        return
```

Comment:
Have `df = None` set at the top of the function, then assign a new value to it as necessary. There should only be one `return` statement at the end of the function, returning the appropriate value of `df`

Reply:
addressed in fd10479